File: ED/Haskell/InFix.hs
module InFix (
 evaluateInfix
 ) where

import Expression
import DataStructures.Stack.LinearStack as S

sample :: Expression
sample = [LeftP, LeftP, Value 4, Mul, Value 5,RightP, Dif, Value 6, RightP]


evaluateInfix :: Expression -> Integer
evaluateInfix expression = evaluateInfix' expression S.empty S.empty

evaluateInfix' :: Expression -> Stack Integer -> Stack Item -> Integer
evaluateInfix' [] enteros operaciones 
    | isEmpty operaciones = S.top enteros
    | otherwise = error "Expresion erroena"
evaluateInfix' (x:expression) enteros operaciones 
    | x == Add || x == Mul || x == Dif = evaluateInfix' expression enteros (S.push x operaciones)
    | x == LeftP = evaluateInfix' expression enteros operaciones
    | x == RightP = evaluateInfix'' (expression) (S.top enteros) (S.top (S.pop enteros)) (S.top operaciones) (S.pop enteros) (S.pop operaciones)  -- evaluateInfix' expression (S.push (operate operador elem2 elem1)) operaciones
    | otherwise = evaluateInfix' expression (S.push (value x) enteros) operaciones
    where
        evaluateInfix'' :: Expression ->  Integer -> Integer -> Item -> Stack Integer -> Stack Item -> Integer
        evaluateInfix'' expression elem1 elem2 operador enteros operaciones = evaluateInfix' expression (S.push (operate operador elem2 elem1) enteros) operaciones


operate :: Item -> Integer -> Integer -> Integer
operate Add x y = x + y
operate Mul x y = x * y
operate Dif x y = x - y


retVal :: Item -> Integer
retVal (Value a) = a

File: ED/relacion3/Java/src/ej3_4/Infix.java
package ej3_4;

import dataStructures.stack.*;

public class Infix {

    static int evaluate(Item[] exprList) {
        Stack<Integer> stackDatas = new ArrayStack<Integer>();
        Stack<Item> stackOperations = new ArrayStack<Item>();
        for (Item expr : exprList) {
            if (expr.isData()) {

                stackDatas.push(expr.getValue());

            } else if (expr.isOperation()) {

                stackOperations.push(expr);

            } else if (expr.isParentheses()) {

                if(expr.isRightP()){

                    int d1 = stackDatas.top();
                    stackDatas.pop();

                    int d2 = stackDatas.top();
                    stackDatas.pop();

                    Item op = stackOperations.top();
                    stackOperations.pop();

                    stackDatas.push(op.evaluate(d2, d1));
                }
            }
        }

        if(stackOperations.isEmpty()){

            return stackDatas.top();

        }else{

            throw new RuntimeException("La expresión no está bien construida");
        }


    }
    public static void main(String [] args) {

        Item [] sample = {
                new LeftP(),
                new LeftP(),
                new Data(4),
                new Mul(),
                new Data(5),
                new RightP(),
                new Dif(),
                new Data(6),
                new RightP()};

        System.out.println(evaluate(sample));
    }
}
